File: src/main/kotlin/leetcode/p0015_3sum/solution.cpp
#include <math.h>
#include <memory.h>
#include <iostream>
#include <algorithm>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
   public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> result;

        int size = nums.size();
        std::unordered_map<int, int> map;
        for (int i = 0; i < size; i++) {
            map[nums[i]] = i;
        }

        for (int i = 0; i < size - 2; i++) {
            if (nums[i] > 0)
                break;

            for (int j = i + 1; j < size - 1; j++) {
                int s = -nums[i] - nums[j];
                auto f = map.find(s);
                if (f != map.end() && f->second > j){
                    result.push_back({nums[i], nums[j], s});
                }
                j = map.find(nums[j])->second;
            }
            i = map.find(nums[i])->second;
        }

        return result;
    }
};

int main() {
    Solution* s = new Solution();

    int size;
    cin >> size;

    vector<int> nums;

    for (size_t i = 0; i < size; i++) {
        int num;
        cin >> num;
        nums.push_back(num);
    }

    vector<vector<int>> res = s->threeSum(nums);
    cout << "Result: \n";
    for (auto a : res) {
        cout << "List: \n";
        for (auto i : a) {
            cout << i << "\n";
        }
    }

    return 0;
}

File: src/main/kotlin/leetcode/p0015_3sum/solution.kt
package leetcode.p0015_3sum

class Solution {
    fun threeSum(nums: IntArray): List<List<Int>> {
        if (nums.size == 3)
            return if (nums.sum() == 0) listOf(nums.toList()) else listOf()

        nums.sort()
        val map = HashMap<Int, Int>()
        nums.forEachIndexed { index, i ->
            map[i] = index
        }

        val result = arrayListOf<List<Int>>()

        var i = 0
        while (i < nums.size - 2){
            if (nums[i] > 0) break

            var j = i + 1
            while (j < nums.size - 1){
                val t = -(nums[i] + nums[j])
                if (map.getOrDefault(t, -1) > j)
                    result.add(listOf(nums[i], nums[j], t))
                j = map[nums[j]]!! + 1
            }

            i = map[nums[i]]!! + 1
        }

        return result
    }
}